controllers/machinepool_controller_phases.go

```diff
@@ -0,0 +1,260 @@
+	capierrors "sigs.k8s.io/cluster-api/errors"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+func (r *MachinePoolReconciler) reconcilePhase(mp *clusterv1.MachinePool) {
+	// Set the phase to "pending" if nil.
+	if mp.Status.Phase == "" {
+		mp.Status.SetTypedPhase(clusterv1.MachinePoolPhasePending)
+	}
+
+	// Set the phase to "provisioning" if bootstrap is ready and the infrastructure isn't.
+	if mp.Status.BootstrapReady && !mp.Status.InfrastructureReady {
+		mp.Status.SetTypedPhase(clusterv1.MachinePoolPhaseProvisioning)
+	}
+
+	// Set the phase to "provisioned" if the infrastructure is ready.
+	if mp.Status.InfrastructureReady {
+		mp.Status.SetTypedPhase(clusterv1.MachinePoolPhaseProvisioned)
+	}
+
+	// Set the phase to "running" if there is a NodeRef field.
+	if mp.Status.InfrastructureReady && len(mp.Status.NodeRefs) == int(mp.Status.ReadyReplicas) {
+		mp.Status.SetTypedPhase(clusterv1.MachinePoolPhaseRunning)
+	}
+
+	// Set the phase to "failed" if any of Status.FailureReason or Status.FailureMessage is not-nil.
+	if mp.Status.FailureReason != nil || mp.Status.FailureMessage != nil {
+		mp.Status.SetTypedPhase(clusterv1.MachinePoolPhaseFailed)
+	}
+
+	// Set the phase to "deleting" if the deletion timestamp is set.
+	if !mp.DeletionTimestamp.IsZero() {
+		mp.Status.SetTypedPhase(clusterv1.MachinePoolPhaseDeleting)
+	}
+}
+
+// reconcileExternal handles generic unstructured objects referenced by a MachinePool.
+func (r *MachinePoolReconciler) reconcileExternal(ctx context.Context, cluster *clusterv1.Cluster, m *clusterv1.MachinePool, ref *corev1.ObjectReference) (external.ReconcileOutput, error) {
+	logger := r.Log.WithValues("machinepool", m.Name, "namespace", m.Namespace)
+
+	obj, err := external.Get(ctx, r.Client, ref, m.Namespace)
+	if err != nil {
+		if apierrors.IsNotFound(errors.Cause(err)) {
+			return external.ReconcileOutput{}, errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: externalReadyWait},
+				"could not find %v %q for MachinePool %q in namespace %q, requeuing",
+				ref.GroupVersionKind(), ref.Name, m.Name, m.Namespace)
+		}
+		return external.ReconcileOutput{}, err
+	}
+
+	// if external ref is paused, return error.
+	if util.IsPaused(cluster, obj) {
+		logger.V(3).Info("External object referenced is paused")
+		return external.ReconcileOutput{Paused: true}, nil
+	}
+
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(obj, r.Client)
+	if err != nil {
+		return external.ReconcileOutput{}, err
+	}
+
+	// Set external object OwnerReference to the MachinePool.
+	ownerRef := metav1.OwnerReference{
+		APIVersion: clusterv1.GroupVersion.String(),
+		Kind:       "MachinePool",
+		Name:       m.Name,
+		UID:        m.UID,
+	}
+
+	// Add ownerRef to object.
+	obj.SetOwnerReferences(util.EnsureOwnerRef(obj.GetOwnerReferences(), ownerRef))
```

Comment:
@vincepri @detiber do you want a controller owner here?

Reply:
Yes, we should set a controllerref here similar to what we are doing in the MachineController's reconcileExternal: https://github.com/kubernetes-sigs/cluster-api/blob/4ec4ec416a91317f4e43cb2d885f592b776072c0/controllers/machine_controller_phases.go#L102-L105